File: scripts/qa/guard_generated_tree.py
#!/usr/bin/env python3
"""Guard against manual edits in generated trees.

Policy:
- Human commits cannot directly modify/add files under:
  - modules/
  - rulesets/
  - dns/
- Pure renames are allowed.
- If generated-tree content changes exist, the same commit must also modify
  at least one file under scripts/ (script-layer source of truth).
- Bot commits are exempt.
"""

from __future__ import annotations

import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
GENERATED_PREFIXES = ("modules/", "rulesets/", "dns/")
SCRIPT_PREFIX = "scripts/"
BOT_AUTHORS = {"github-actions[bot]"}


def sh(*args: str) -> str:
    p = subprocess.run(args, cwd=ROOT, capture_output=True, text=True, check=True)
    return p.stdout.strip()


def main() -> int:
    base = sh("git", "rev-parse", "HEAD~1")
    head = sh("git", "rev-parse", "HEAD")
    author = sh("git", "log", "-1", "--pretty=%an")

    if author in BOT_AUTHORS:
        print(f"✅ guard: bot commit ({author}) exempt.")
        return 0

    rows = sh("git", "diff", "--name-status", base, head).splitlines()
    if not rows:
        print("✅ guard: no changes.")
        return 0

    changed_scripts = False
    generated_content_changes: list[str] = []

    for row in rows:
        parts = row.split("\t")
        if not parts:
            continue
        status = parts[0]
        paths = parts[1:]
        for p in paths:
            if p.startswith(SCRIPT_PREFIX):
                changed_scripts = True

        # Rename/copy status uses old+new paths. Allow pure renames by policy.
        if status.startswith(("R", "C")):
            continue

        target = paths[-1] if paths else ""
        if target.startswith(GENERATED_PREFIXES):
            generated_content_changes.append(f"{status}\t{target}")

    if generated_content_changes and not changed_scripts:
        print("❌ guard: generated tree content changed without script-layer changes.")
        print("Please modify generation scripts under scripts/ and regenerate outputs.")
        for item in generated_content_changes:
            print(f"  - {item}")
        return 1

    print("✅ guard: generated-tree policy passed.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: go_scripts/pkg/qa/guard_generated_tree.go
package qa

import (
	"fmt"
	"os/exec"
	"strings"

	"github.com/nyamiiko/script_hub/go_scripts/pkg/hub"
)

func runGitCmd(args ...string) string {
	cmd := exec.Command("git", args...)
	cmd.Dir = hub.ROOT
	out, err := cmd.Output()
	if err != nil {
		// return empty string on error, same as python throwing but we just swallow for simplicity
		// or return empty and let caller handle it.
		return ""
	}
	return strings.TrimSpace(string(out))
}

func RunGuardGeneratedTree() int {
	base := runGitCmd("rev-parse", "HEAD~1")
	head := runGitCmd("rev-parse", "HEAD")
	author := runGitCmd("log", "-1", "--pretty=%an")

	if author == "github-actions[bot]" {
		fmt.Printf("✅ guard: bot commit (%s) exempt.\n", author)
		return 0
	}

	diffOut := runGitCmd("diff", "--name-status", base, head)
	if diffOut == "" {
		fmt.Println("✅ guard: no changes.")
		return 0
	}

	rows := strings.Split(diffOut, "\n")
	changedScripts := false
	var generatedContentChanges []string

	validSources := []string{"scripts/", "rulesets/Sources/", "go_scripts/"}
	generatedPrefixes := []string{"modules/", "rulesets/", "dns/"}

	for _, row := range rows {
		parts := strings.Split(row, "\t")
		if len(parts) == 0 {
			continue
		}
		status := parts[0]
		paths := parts[1:]
		for _, p := range paths {
			for _, prefix := range validSources {
				if strings.HasPrefix(p, prefix) {
					changedScripts = true
					break
				}
			}
		}

		if strings.HasPrefix(status, "R") || strings.HasPrefix(status, "C") {
			continue
		}

		target := ""
		if len(paths) > 0 {
			target = paths[len(paths)-1]
		}

		isGenerated := false
		for _, prefix := range generatedPrefixes {
			if strings.HasPrefix(target, prefix) {
				isGenerated = true
				break
			}
		}

		if isGenerated && !strings.HasPrefix(target, "rulesets/Sources/") {
			generatedContentChanges = append(generatedContentChanges, fmt.Sprintf("%s\t%s", status, target))
		}
	}

	if len(generatedContentChanges) > 0 && !changedScripts {
		fmt.Println("❌ guard: generated tree content changed without script-layer changes.")
		fmt.Println("Please modify generation scripts under scripts/ or go_scripts/ and regenerate outputs.")
		for _, item := range generatedContentChanges {
			fmt.Printf("  - %s\n", item)
		}
		return 1
	}

	fmt.Println("✅ guard: generated-tree policy passed.")
	return 0
}
